jax/_src/numpy/polynomials/chebyshev.py

```diff
@@ -0,0 +1,605 @@
+from functools import partial
+import functools
+import operator
+
+import jax.numpy as jnp
+import jax.lax as lax
+from jax import jit
+from jax.numpy.linalg import eigvals
+from jax._src.numpy.util import _promote_dtypes_inexact
+from jax._src.numpy.polynomials import polyutils as pu
+
+__all__ = [
+    'chebzero', 'chebone', 'chebx', 'chebdomain', 'chebline', 'chebadd',
+    'chebsub', 'chebmulx', 'chebmul', 'chebdiv', 'chebpow', 'chebval',
+    'chebder', 'chebint', 'cheb2poly', 'poly2cheb', 'chebfromroots',
+    'chebvander', 'chebfit', 'chebtrim', 'chebroots', 'chebpts1',
+    'chebpts2', 'chebval2d', 'chebval3d', 'chebgrid2d',
+    'chebgrid3d', 'chebvander2d', 'chebvander3d', 'chebcompanion',
+    'chebgauss', 'chebweight', 'chebinterpolate'
+]
+
+#Not compatible with JIT
+chebtrim = pu.trimcoef
+
+@jit
+def _cseries_to_zseries(c):
+    n = c.size
+    zs = jnp.zeros(2*n-1, dtype=c.dtype)
+    zs = zs.at[n-1:].set(c/2)
+    return zs + zs[::-1]
+
+@jit
+def _zseries_to_cseries(zs):
+    n = (zs.size + 1)//2
+    c = zs[n-1:].copy()
+    c = c.at[1:n].multiply(2)
+    return c
+
+@jit
+def _zseries_mul(z1, z2):
+    return jnp.convolve(z1, z2)
+
+@jit
+def _zseries_div(z1, z2):
+    z1 = z1.copy()
+    z2 = z2.copy()
+    lc1 = len(z1)
+    lc2 = len(z2)
+
+    if lc2 == 1:
+        z1 /= z2
+        return z1, z1[:1]*0
+    elif lc1 < lc2:
+        return z1[:1]*0, z1
+    else:
+        dlen = lc1 - lc2
+        scl = z2[0]
+        z2 /= scl
+        quo = jnp.empty(dlen + 1, dtype=z1.dtype)
+        i = 0
+        j = dlen
+
+        while i < j:
+            r = z1[i]
+            quo = quo.at[i].set(z1[j])
+            quo = quo.at[dlen - i].set(r)
+            tmp = r*z2
+            z1 = z1.at[i:i+lc2].add(-tmp)
+            z1 = z1.at[j:j+lc2].add(-tmp)
+            i += 1
+            j -= 1
+        r = z1[i]
+        quo = quo.at[i].set(r)
+        tmp = r*z2
+        z1 = z1.at[i:i+lc2].add(-tmp)
+        quo /= scl
+        rem = z1[i+1:i-1+lc2].copy()
+        return quo, rem
+
+@jit
+def _zseries_der(zs):
+    n = len(zs)//2
+    ns = jnp.array([-1, 0, 1], dtype=zs.dtype)
+    zs *= jnp.arange(-n, n+1)*2
+    d, r = _zseries_div(zs, ns)
+    return d
+
+@jit
+def _zseries_int(zs):
+    n = 1 + len(zs)//2
+    ns = jnp.array([-1, 0, 1], dtype=zs.dtype)
+    zs = _zseries_mul(zs, ns)
+    div = jnp.arange(-n, n+1)*2
+    zs = zs.at[:n].divide(div[:n])
+    zs = zs.at[n+1:].divide(div[n+1:])
+    zs = zs.at[n].set(0)
+    return zs
+
+@jit
+def poly2cheb(pol):
+    [pol] = pu.as_series([pol])
+    deg = len(pol) - 1
+    res = 0
+    for i in range(deg, -1, -1):
+        res = chebadd(chebmulx(res), pol[i])
+    return res[:-1]
+
+@jit
+def cheb2poly(c):
+    [c] = pu.as_series([c])
+    n = len(c)
+    if n < 3:
+        return c
+    else:
+        c0 = jnp.asarray(c[-2])
+        c1 = jnp.asarray(c[-1])
+
+        for i in range(n-1, 1, -1):
```

Comment:
Same comment here about Python control flow when the size gets large. We may want to use `lax.fori_loop` instead unless `n` is bounded by a small number.